# Projeto dedicado a Tokenização
## Project for Tokenization

Neste projeto será implementado um tokenizador fazendo uso do algoritmo BPE para fins de estudo.

That project will implement a tokenizer using the BPE algorithm for study purposes.

Para gerar o vocabulário do tokenizador será preciso importar as instruções básicas do nosso tokenizador e utilizar os arquivos .json que foram reunidos.

To make the vocabulary of the tokenizer will need to import the basic instructions of our tokenizer and utilize some reunited .json files.

In [1]:
import tokenizer as t

Com todas as funções disponíveis, já é possível implementar a construção do vocabulário. Para fazer isso, o código será baseado no exemplo deixado no tokenizer.py.

With all these functions, it's possible to implement vocabulary construction. To start, the code will be based on the example displayed in file tokenizer.py.

In [21]:
import os, json

path = 'corpus/'
files = [file for file in os.listdir(path) if file.endswith('.json')]

merges = {}

# Loop to take all the text in training 
for i in range(len(files)): 
  # Open the JSON file
  f = open(path + files[i], "r")
  data = json.load(f)["text"]
  f.close()

  # Making Vocab
  tokens = list(map(int, data.encode("utf-8")))
  if merges:
    merge, ids = t.makeVocab(261, tokens, merges[list(merges)[-1]] - 256, merges)
  else: 
    merge, ids = t.makeVocab(261, tokens)

  # Keep dict Ordened
  merges = dict(sorted({**merge, **merges}.items(), key=lambda x:x[1]))

print(merges)


{(97, 32): 256, (101, 32): 257, (111, 32): 258, (115, 32): 259, (110, 116): 260, (100, 257): 261, (42, 32): 262, (44, 32): 263, (101, 260): 264, (195, 163): 265, (101, 115): 266, (97, 259): 267, (265, 258): 268, (109, 32): 269, (97, 114): 270, (111, 259): 271, (100, 256): 272, (111, 114): 273, (100, 258): 274, (105, 99): 275, (114, 105): 276, (116, 101): 277, (97, 108): 278, (195, 161): 279, (83, 111): 282, (101, 110): 286, (32, 261): 288, (97, 116): 290, (116, 105): 292, (101, 114): 295, (105, 110): 296, (32, 112): 297, (97, 115): 298, (32, 111): 299, (32, 42): 300, (32, 49): 301, (100, 101): 302, (48, 49): 303, (298, 32): 305, (32, 100): 305, (305, 101): 309, (97, 110): 313, (313, 116): 315, (97, 103): 316, (83, 315): 317, (97, 100): 319, (300, 262): 321, (32, 262): 322, (105, 114): 324, (105, 97): 325, (256, 100): 329, (114, 97): 333, (114, 101): 337, (46, 32): 338, (115, 116): 341, (111, 110): 342, (195, 169): 346, (100, 97): 347, (100, 111): 349, (32, 67): 349, (99, 97): 350, (103

Com o vocabulário que foi produzido é possível codificar e decodificar qualquer texto usando as funções que já foram implementadas. Segue exemplo a seguir:

Using the obtained vocabulary it's possible to code and decode any text using the functions implemented before. Follow an example: 

In [22]:
# Take te string to be tokenized
text = "Some random string to be used as an example to be tokenized. To have the best use of the tokenizer and have a good result, it is important to have some long enough text to find some pattern to be followed. This example is not good but it is some start to understand the proposal and the ideas behind the code being done.\nExpanding the text!\nOn the first try, this text was long, but the number of tokens was exactly equal to the number of Unicode bytes, for this reason was increased a little bit the size of the text (probably will not change the equality of the sizes but will provide a better result in the end)."

tokens = list(map(int, text.encode("utf-8")))
# Creating dictionary with all vocab
vocab = {idx: bytes([idx]) for idx in range(256)}
# This part only works because we have sure that merges indexes are in increasing order
for (p0, p1), idx in merges.items():
  vocab[idx] = vocab[p0] + vocab[p1]

# Results
print("Text length: ", len(text))
print("Initial tokens length: ", len(tokens))
print("Final tokens length: ", len(t.encoder(text, merges)))
print(f"Compression rate: {len(tokens)/len(t.encoder(text, merges)):.2f}X")

print(t.decoder(t.encoder(text, merges), vocab))

Text length:  615
Initial tokens length:  615
Final tokens length:  371
Compression rate: 1.66X
Some ran Cm string to be used as an example to be tokenized. To have the best use of the tokenizer and have a good result, it is important to have s::e long enough text to find s::e pattern to be followed. This example is not good but it is s::e start to understand the proposal and the ideas behind the code being done.
Expanding the text!
On the first try, this text was long, but the number of tokens was exactly equal to the number of Unicode bytes, for this reason was increased a little bit the size of the text (probably wil13not change the equality of the sizes but will provide a better result in the end).


OBS: O tamanho do vocabulário do treinamento como 261 (5 a mais do que os bytes definidos pelo Unicode) foi escolhido por gerar um vocabulário final reduzido, mas mantendo um bom índice de compactação. Foi testado com 266 e 276 obtendo compactação de 1.86 e 2.09, respectivamente.

OBS: The size of the training vocabulary, 261 (5 more than the default size), was chosen to reduce the final vocabulary while maintaining a good rate of compression. It was tested with 266 and 276, getting a compression of 1.86 and 2.09, respectively.  